pkg/operator/ceph/cluster/controller.go

```diff
@@ -492,24 +513,52 @@ func (c *ClusterController) onUpdate(oldObj, newObj interface{}) {
 
 	cluster.Spec = &newClust.Spec
 
-	// attempt to update the cluster.  note this is done outside of wait.Poll because that function
-	// will wait for the retry interval before trying for the first time.
-	done, _ := c.handleUpdate(newClust.Name, cluster)
-	if done {
-		return
-	}
+	// Launch handleUpdate in a goroutine. If an existing goroutine is found,
+	// stop it and wait for it to exit before launching the goroutine
 
-	err = wait.Poll(updateClusterInterval, updateClusterTimeout, func() (bool, error) {
-		return c.handleUpdate(newClust.Name, cluster)
-	})
-	if err != nil {
-		message := fmt.Sprintf("giving up trying to update cluster in namespace %s after %s", cluster.Namespace, updateClusterTimeout)
-		logger.Error(message)
-		if err := c.updateClusterStatus(newClust.Namespace, newClust.Name, cephv1.ClusterStateError, message); err != nil {
-			logger.Errorf("failed to update cluster status in namespace %s: %+v", newClust.Namespace, err)
+	if ctx, ok := c.updaterCtxMap[newClust.Name]; ok {
+		select {
+		case <-ctx.Done():
+			// nothing to do if previous updater has finished
+		default:
+			ctx.cancel()
+			<-ctx.wait
 		}
-		return
 	}
+
+	ctx := newUpdaterCtx()
+	c.updaterCtxMap[newClust.Name] = ctx
+
+	go func() {
+		err = wait.PollImmediate(updateClusterInterval, updateClusterTimeout, func() (bool, error) {
```

Comment:
Should we get rid of the call to `wait.PollImmediate()` and add another case to the select statement for the `updateClusterInterval`? This way the `Done()` case can be triggered before the 30s timeout for the wait.PollImmediate.

Reply:
We can use `wait.UntilWithContext` and pass a context with a Timeout. That should cancel immediately.